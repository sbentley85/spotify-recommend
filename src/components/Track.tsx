import React from "react";

const Track = (props: {
	track: { id: string; name: string; artist: string };
	addToPicks: any;
}) => {
	return (
		<div onClick={(e) => props.addToPicks(e, props.track.id)}>
			{props.track.artist + " - " + props.track.name}
		</div>
	);
};

export default Track;
