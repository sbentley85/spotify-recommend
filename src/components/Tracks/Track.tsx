import React, { SyntheticEvent, useState } from "react";
import { IPicks } from "../Choices/Choices";
import trackStyles from "./tracks.module.css";
import { ITrackClick } from "../Recommendations/MyRecommendations";
import Likebutton from "./Likebutton";
import SpotifyUtils from "../../util/spotify";

const Track = (props: { track: IPicks; handleClick: ITrackClick }) => {
	const [liked, setLiked] = useState<boolean>(props.track.liked!);

	const handleLike = (event: SyntheticEvent) => {
		// send request to spotify, returns true is request successful
		// if successful update liked state variable
		if (SpotifyUtils.likeTrack(props.track.id)) setLiked(true);
	};

	const handleUnLike = (event: SyntheticEvent) => {
		// send request to spotify, returns true is request successful
		// if successful update liked state variable
		if (SpotifyUtils.likeTrack(props.track.id)) setLiked(false);
	};

	return (
		<div className={trackStyles.track}>
			<div className={trackStyles.trackWrapper}>
				<div
					className={trackStyles.trackImage}
					style={
						props.track.smImg
							? {
									backgroundImage: `url(${props.track.smImg.url})`,
							  }
							: { backgroundColor: "black" }
					}
				></div>
				<div
					className={trackStyles.trackDetails}
					onClick={(e) => props.handleClick(e, props.track)}
				>
					<div className={trackStyles.trackDetail}>
						{props.track.artist}

						{"  -  "}

						{props.track.name}
					</div>
				</div>
			</div>

			<Likebutton
				liked={liked!}
				handleLike={handleLike}
				handleUnLike={handleUnLike}
			/>
		</div>
	);
};

export default Track;
