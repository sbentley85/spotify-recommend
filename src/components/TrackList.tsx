import React from "react";
import Track from "./Track";
import { IPicks } from "./Choices";

const TrackList = (props: { handleClick?: any; tracks: IPicks[] }) => {
	return props.tracks.length !== 0 ? (
		<div className="trackList">
			{props.tracks.map((track: IPicks, index: number) => {
				return (
					<Track
						handleClick={props.handleClick}
						track={track}
						key={index}
					/>
				);
			})}
		</div>
	) : null;
};

export default TrackList;
