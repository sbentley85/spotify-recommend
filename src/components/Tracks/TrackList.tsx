import React from "react";
import Track from "./Track";
import { IPicks } from "../Choices/Choices";
import trackStyles from "./tracks.module.css";
import { ITrackClick } from "../Recommendations/MyRecommendations";

const TrackList = (props: { handleClick: ITrackClick; tracks: IPicks[] }) => {
	return props.tracks.length !== 0 ? (
		<div className={trackStyles.trackList}>
			{props.tracks.map((track: IPicks, index: number) => {
				return (
					<Track
						handleClick={props.handleClick}
						track={track}
						key={index}
					/>
				);
			})}
		</div>
	) : null;
};

export default TrackList;
