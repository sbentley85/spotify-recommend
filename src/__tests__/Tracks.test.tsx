import React from "react";
import Track from "../components/Tracks/Track";
import { shallow } from "enzyme";
import TrackList from "../components/Tracks/TrackList";
import trackStyles from "./tracks.module.css";

// sample track
const sampleTrack = {
	id: "69u375BY4upymJzyAnSXRN",
	name: "You Caught the Light",
	artist: "CHVRCHES",
	medImg: {
		height: 300,
		url: "https://i.scdn.co/image/ab67616d00001e027ecffdcb0c079ce32b47be74",
		width: 300,
	},
	smImg: {
		height: 64,
		url: "https://i.scdn.co/image/ab67616d000048517ecffdcb0c079ce32b47be74",
		width: 64,
	},
	uri: "spotify:track:69u375BY4upymJzyAnSXRN",
};

const sampleTracks = [
	{
		id: "69u375BY4upymJzyAnSXRN",
		name: "You Caught the Light",
		artist: "CHVRCHES",
		medImg: {
			height: 300,
			url:
				"https://i.scdn.co/image/ab67616d00001e027ecffdcb0c079ce32b47be74",
			width: 300,
		},
		smImg: {
			height: 64,
			url:
				"https://i.scdn.co/image/ab67616d000048517ecffdcb0c079ce32b47be74",
			width: 64,
		},
		uri: "spotify:track:69u375BY4upymJzyAnSXRN",
	},
	{
		id: "3rtJW3GqUTBJXF9nAv3Bp6",
		name: "Cosmonaut",
		artist: "At The Drive In",
		medImg: {
			height: 300,
			url:
				"https://i.scdn.co/image/ab67616d00001e02d8077764c1d0f76baac75478",
			width: 300,
		},
		smImg: {
			height: 64,
			url:
				"https://i.scdn.co/image/ab67616d00004851d8077764c1d0f76baac75478",
			width: 64,
		},
		uri: "spotify:track:3rtJW3GqUTBJXF9nAv3Bp6",
	},
];

describe("Track component", () => {
	it("renders Track component without crashing", () => {
		const mytrack = shallow(
			<Track
				track={sampleTrack}
				handleClick={() => {
					console.log("testing Track component onclick");
				}}
			/>
		);
	});

	it("renders Track & Artist name from props", () => {
		const myTrack = shallow(
			<Track
				track={sampleTrack}
				handleClick={() => {
					console.log("testing Track component onclick");
				}}
			/>
		);
		expect(myTrack).toMatchSnapshot();
	});

	const clickFn = jest.fn();

	// onclick moved temporarily to trackwrapper div
	it("track click should call onClick function", () => {
		const myTrack = shallow(
			<Track track={sampleTrack} handleClick={clickFn} />
		);
		const trackWrapperDiv = myTrack.find(`.${trackStyles.trackDetails}`);
		trackWrapperDiv.simulate("click");
		expect(clickFn).toHaveBeenCalled();
	});
});

describe("TrackList component", () => {
	const clickFn = jest.fn();
	it("renders Tracklist without crashing", () => {
		shallow(<TrackList tracks={sampleTracks} handleClick={clickFn} />);
	});

	it("renders a list of tracks from props", () => {
		const myTrackList = shallow(
			<TrackList tracks={sampleTracks} handleClick={clickFn} />
		);
		expect(myTrackList).toMatchSnapshot();
	});
});
