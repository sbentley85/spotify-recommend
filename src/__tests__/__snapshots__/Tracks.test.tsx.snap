// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track component renders Track & Artist name from props 1`] = `
<div
  className="track"
>
  <div
    className="trackWrapper"
  >
    <div
      className="trackImage"
      style={
        Object {
          "backgroundImage": "url(https://i.scdn.co/image/ab67616d000048517ecffdcb0c079ce32b47be74)",
        }
      }
    />
    <div
      className="trackDetails"
      onClick={[Function]}
    >
      <div
        className="trackDetail"
      >
        CHVRCHES
          -  
        You Caught the Light
      </div>
    </div>
  </div>
  <Likebutton
    handleLike={[Function]}
    handleUnLike={[Function]}
  />
</div>
`;

exports[`TrackList component renders a list of tracks from props 1`] = `
<div
  className="trackList"
>
  <Track
    handleClick={[MockFunction]}
    key="0"
    track={
      Object {
        "artist": "CHVRCHES",
        "id": "69u375BY4upymJzyAnSXRN",
        "medImg": Object {
          "height": 300,
          "url": "https://i.scdn.co/image/ab67616d00001e027ecffdcb0c079ce32b47be74",
          "width": 300,
        },
        "name": "You Caught the Light",
        "smImg": Object {
          "height": 64,
          "url": "https://i.scdn.co/image/ab67616d000048517ecffdcb0c079ce32b47be74",
          "width": 64,
        },
        "uri": "spotify:track:69u375BY4upymJzyAnSXRN",
      }
    }
  />
  <Track
    handleClick={[MockFunction]}
    key="1"
    track={
      Object {
        "artist": "At The Drive In",
        "id": "3rtJW3GqUTBJXF9nAv3Bp6",
        "medImg": Object {
          "height": 300,
          "url": "https://i.scdn.co/image/ab67616d00001e02d8077764c1d0f76baac75478",
          "width": 300,
        },
        "name": "Cosmonaut",
        "smImg": Object {
          "height": 64,
          "url": "https://i.scdn.co/image/ab67616d00004851d8077764c1d0f76baac75478",
          "width": 64,
        },
        "uri": "spotify:track:3rtJW3GqUTBJXF9nAv3Bp6",
      }
    }
  />
</div>
`;
